### ISSUE TITLE
[RFC]: How should we emulate methods?

### ISSUE THREAD

**Issue by mhevery** [BuilderIO/qwik#617]
### Is your feature request related to a problem?

u

### Describe the solution you'd like

u

### Describe alternatives you've considered

u

**Comment by mahmoudajawad** [comment]
@mhevery, why does Qwik have to implement such a feature? Any use case that prompts such a need?

**Comment by mhevery** [comment]
The motivation comes from here: https://github.com/BuilderIO/builder/blob/main/packages/sdks/src/blocks/columns/columns.lite.tsx

It is totally reasonable that states have getters/setters as well as behavior functions. The issue is that we don't know how to serialize the behavior in the state. And even if we could it is not clear at what point it should be recovered (tree shaking, maybe the method is only needed in a listener, so restoring it as part of render is wasteful)

The way I have solved the above example is that I have created top-level functions that take `self` as the first argument. The advantage of that approach is that it is tree shakable. The downside is that you don't have an OO feel and you can't take advantage of polymorphism. 

I think the above solution is good enough for now.

**Comment by steve8708** [comment]
I also agree with this not being needed https://github.com/BuilderIO/builder/pull/1081#discussion_r902750134

**Comment by mhevery** [comment]
Let me provide a simpler example:
```typescript
export const MyComponent = component$(() => {
  const store = useStore({count:0});

  function increment() {
    store.count++;
  }

  return <button onClick$={() => increment(store)}>{store.count}</button>
});
```

Here I would like to encapsulate the `increment` business logic example. The problem is as written above it will not work because `incemente` needs to cross `$` boundary.

So the question is How can behavior be added to stores in a way so that it can be used from any where????

Possible solution:
```typescript
export const MyComponent = component$(() => {
  const store = useStore({count:0}, $((store) => {
    store.increment = () => this.count++;
  }));

  return <button onClick$={() => increment(store)}>{store.count}</button>
});
```

The above would work, but would require some changes to the framework.

**Comment by akaufmann** [comment]
First of all, a great job on this innovative framework to solve one of "the" problems of current websites.

+1 from my side. This is a disadvantage of the very innovative Qwik, that you can't use (normal) functions inside event callbacks, especially if you have to provide parameters. Either you have to put every logic in the cb function, which gets messy very fast with JSX if you have complex logic or you have to use QRLs that you have to invoke. I know why this is, but I'm looking forward to a smarter way to handle these scenarios w/o seeing -> `Code(3): Only primitive and object literals can be serialized (foo: Foo)=>{\n ...\n }`.

**Comment by mhevery** [comment]
Actually this would work today, but not very nice DX:

```typscript
export const MyComponent = component$(() => {
  const store = useStore({count:0});

  const increment = $(function() {
    store.count++;
  });

  return <button onClick$={async () => await increment.invoke()}>{store.count}</button>
});
```

**Comment by mhevery** [comment]
Well no. Problem is that the calls it would have to be translated form sync to async and that is no trivial. For example `foo.id` would now return a promise instead of value.

For now we are not blocked on this but it is something to think about.

**Comment by steve8708** [comment]
ah yeah I see, so what we really want is something like vscode's "extract to top level function" functionality, where we can pull it to be top level, add export keyword, and pass all references used as arguments

**Comment by mhevery** [comment]
Which is exactly what I have implemented for now.

But long term we went something better.

**Comment by steve8708** [comment]
If I'm not crazy this actually is already supported? cc @manucorporat maybe can clarify 

<img width="1163" alt="image" src="https://user-images.githubusercontent.com/844291/175422300-5187d6ed-7874-40cc-9a39-b114d7891aa8.png">


[Playground link](https://qwik.builder.io/playground#version=0.0.30&buildMode=development&entryStrategy=hook&files=JTVCJTdCJTIycGF0aCUyMiUzQSUyMiUyRmFwcC50c3glMjIlMkMlMjJjb2RlJTIyJTNBJTIyaW1wb3J0JTIwJTdCJTIwY29tcG9uZW50JTI0JTJDJTIwdXNlU3RvcmUlMjAlN0QlMjBmcm9tJTIwJyU0MGJ1aWxkZXIuaW8lMkZxd2lrJyUzQiU1Q24lNUNuZXhwb3J0JTIwY29uc3QlMjBBcHAlMjAlM0QlMjBjb21wb25lbnQlMjQoKCklMjAlM0QlM0UlMjAlN0IlNUNuJTIwJTIwY29uc3QlMjBzdGF0ZSUyMCUzRCUyMHVzZVN0b3JlKCU3QiUyMG5hbWUlM0ElMjAnU3RldmUnJTIwJTdEKSU1Q24lNUNuJTIwJTIwZnVuY3Rpb24lMjBmb28oYXJnKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjByZXR1cm4lMjBzdGF0ZS5uYW1lJTIwJTJCJTIwYXJnJTVDbiUyMCUyMCU3RCU1Q24lNUNuJTIwJTIwcmV0dXJuJTIwJTNDcCUzRUNvb2wlMjBRd2lrJTIwJTdCZm9vKCclMjAnKSU3RCUzQyUyRnAlM0UlM0IlNUNuJTdEKSUzQiU1Q24lMjIlN0QlMkMlN0IlMjJwYXRoJTIyJTNBJTIyJTJGZW50cnkuc2VydmVyLnRzeCUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjAlN0IlMjByZW5kZXJUb1N0cmluZyUyQyUyMFJlbmRlck9wdGlvbnMlMjAlN0QlMjBmcm9tJTIwJyU0MGJ1aWxkZXIuaW8lMkZxd2lrJTJGc2VydmVyJyUzQiU1Q25pbXBvcnQlMjAlN0IlMjBSb290JTIwJTdEJTIwZnJvbSUyMCcuJTJGcm9vdCclM0IlNUNuJTVDbmV4cG9ydCUyMGZ1bmN0aW9uJTIwcmVuZGVyKG9wdHMlM0ElMjBSZW5kZXJPcHRpb25zKSUyMCU3QiU1Q24lMjAlMjByZXR1cm4lMjByZW5kZXJUb1N0cmluZyglM0NSb290JTIwJTJGJTNFJTJDJTIwb3B0cyklM0IlNUNuJTdEJTVDbiUyMiU3RCUyQyU3QiUyMnBhdGglMjIlM0ElMjIlMkZyb290LnRzeCUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjAlN0IlMjBBcHAlMjAlN0QlMjBmcm9tJTIwJy4lMkZhcHAnJTNCJTVDbiU1Q25leHBvcnQlMjBjb25zdCUyMFJvb3QlMjAlM0QlMjAoKSUyMCUzRCUzRSUyMCU3QiU1Q24lMjAlMjByZXR1cm4lMjAoJTVDbiUyMCUyMCUyMCUyMCUzQ2h0bWwlM0UlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTNDaGVhZCUzRSU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlM0N0aXRsZSUzRUhlbGxvJTIwUXdpayUzQyUyRnRpdGxlJTNFJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUzQyUyRmhlYWQlM0UlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTNDYm9keSUzRSU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlM0NBcHAlMjAlMkYlM0UlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTNDJTJGYm9keSUzRSU1Q24lMjAlMjAlMjAlMjAlM0MlMkZodG1sJTNFJTVDbiUyMCUyMCklM0IlNUNuJTdEJTNCJTVDbiUyMiU3RCU1RA%3D%3D)

**Comment by mahmoudajawad** [comment]
If I recall correctly, I also had something like this used in 0.0.2[45] and it worked.

**Comment by mahmoudajawad** [comment]
So, I just upgraded one test project to 0.0.30 and now the following would break the app:
```jsx
export const ROUTES: Record<string, FunctionComponent> = {
  "/": () => <HomePage greeting="Home!" />,
  "/about": () => <AboutPage greeting="About!" />,       
};

export const App = component$(({ opts }: { opts: RenderToStringOptions }) => {
const state = useStore({router: ROUTES['/'],});
// ...
});
```

If I down-grade to 0.0.26 it works again, so wondering what have changed in these 4 versions.

**Comment by mahmoudajawad** [comment]
@manucorporat, can I hear your input on what I mentioned in [previous comment](https://github.com/BuilderIO/qwik/issues/617#issuecomment-1165819271)? It's 0.0.34 and the behaviour still differs from how it was with 0.0.29 when it worked. Is this the state going forward? Why did it work before 0.0.30 and what is the reason it is not now?
